src/sql/reuse/relaxed.rs

```diff
@@ -0,0 +1,147 @@
+use sql::reuse::helpers::predicate_implication::complex_predicate_implies;
+use sql::reuse::{ReuseConfiguration, ReuseType};
+use sql::query_graph::{QueryGraph, QueryGraphEdge};
+use mir::MirQuery;
+
+use std::vec::Vec;
+use std::collections::HashMap;
+
+
+/// Implementation of reuse algorithm with relaxed constraints.
+/// While Finkelstein checks if queries are compatible for direct extension,
+/// this algorithm considers the possibility of reuse of internal views.
+/// For example, given the queries:
+/// 1) select * from Paper, PaperReview where Paper.paperId = PaperReview.paperId and PaperReview.reviewType = 1;
+/// 2) select * from Paper, PaperReview where Paper.paperId = PaperReview.paperId;
+///
+/// Finkelstein reuse would be conditional on the order the queries are added,
+/// since 1) is a direct extension of 2), but not the other way around.
+///
+/// This relaxed version of the reuse algorithm considers cases where a query might
+/// reuse just a prefix of another. First, it checks if the queries perform the
+/// same joins, then it checks predicate implication and at last, checks group
+/// by compatibility.
+/// If all checks pass, them the algorithm works like Finkelstein and the query
+/// is a direct extension of the other. However, if not all, but at least one
+/// check passes, then the algorithm returns that the queries have a prefix in
+/// common.
+pub struct Relaxed;
+
+impl ReuseConfiguration for Relaxed {
+    fn reuse_candidates<'a>(qg: &QueryGraph, query_graphs: &'a HashMap<u64, (QueryGraph, MirQuery)>) -> Vec<(ReuseType, &'a QueryGraph)>{
+        let mut reuse_candidates = Vec::new();
+        for &(ref existing_qg, _) in query_graphs.values() {
+            if existing_qg
+                .signature()
+                .is_weak_generalization_of(&qg.signature())
+            {
+                match Self::check_compatibility(&qg, existing_qg) {
+                    Some(reuse) => {
+                        // QGs are compatible, we can reuse `existing_qg` as part of `qg`!
+                        reuse_candidates.push((reuse, existing_qg));
+                    }
+                    None => (),
+                }
+            }
+        }
+
+        reuse_candidates
+    }
+}
+
+impl Relaxed {
+    fn check_compatibility(new_qg: &QueryGraph, existing_qg: &QueryGraph) -> Option<ReuseType> {
+        // 1. NQG's nodes is subset of EQG's nodes
+        // -- already established via signature check
+        assert!(
+            existing_qg
+                .signature()
+                .is_weak_generalization_of(&new_qg.signature())
+        );
+
+        // Check if the queries are join compatible -- if the new query
+        // performs a superset of the joins in the existing query.
```

Comment:
Hmm, here's something we might need to be aware of: this code no longer checks for compatibility of join predicates, only that a join between the same two nodes exists. This is correct as long as we only allow equi-joins and as long as join predicates are simple (i.e., no additional ANDs and ORs within them), which I believe is currently always the case.

The old true-to-Finkelstein implementation guaranteed this because it checked all the predicates _first_ before looking at the join edges; here, we rely on the MIR-level reuse algorithm to check correctness of the join columns (which it indeed does).

In other words, no problem as such, but we need to be careful if we allow more powerful joins later (e.g., HotCRP-style `a JOIN b ON (a.c = b.d AND a.c != <something>)`).

Reply:
Ah, yeah. Ideally, we would also check predicate implication in the the join predicate. Though, I don't think the Finkelstein algorithm actually handles this. It only checks predicate implication in non-edge predicates.